### DISCUSSION TITLE
CH32V103 Support

### DISCUSSION THREAD

**Discussion by ProxyPlayerHD** [hathach/tinyusb#2605] [Ideas]
the CH32V103 is the cheapest 5V MCU i know with a USB host/device controller. so it would be amazing if tinyUSB supported it. it's similar to the CH32V20x/307, but there are some slight differences with the USB controller.

i briefly tried myself at adding host support for it a while ago, but ran into some issues.
specifically i didn't realize that the 20x/307 ports only implement the device API, not the host API (which is the one i wanted). plus i wasn't able to write the host API port from scratch either because there appears to be no documentation on it (the [porting guide](https://docs.tinyusb.org/en/latest/contributing/porting.html) only explains the device API)

so beyond having a port, some updated documentation would also be nice...

**Comment by mikeemoo** [comment]
+1 for this. I'd really like to be able to my ch32v203 as host and device at the same time.

**Comment by UMRnInside** [comment]
(late) update: CH32V103 gets (partial) device mode support & Hardware-in-the-Loop test coverage in TinyUSB 0.21.0. You can build & flash device mode examples, and see (most of) them working on real CH32V103 MCUs.
https://github.com/hathach/tinyusb/pull/3710 https://github.com/hathach/tinyusb/pull/3620
